File: Classes/Model classes/Airports.m
/*
	MyFlightbook for iOS - provides native access to MyFlightbook
	pilot's logbook
 Copyright (C) 2017-2021 MyFlightbook, LLC
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//
//  Airports.m
//  MFBSample
//
//  Created by Eric Berman on 12/25/09.
//  Copyright 2009-2021, MyFlightbook LLC. All rights reserved.
//

#import "Airports.h"
#import "LocalAirports.h"
#import "MFBAppDelegate.h"

@implementation Airports

#define MIN_NAVAID_CODE_LENGTH  2
#define MIN_AIRPORT_CODE_LENGTH 3
#define MAX_AIRPORT_CODE_LENGTH 6

#define szNavaidPrefix @"@"

@synthesize errorString;
@synthesize rgAirports;

+ (NSString *) RegAdHocFix
{
    return [NSString stringWithFormat:@"%@%@", szNavaidPrefix, @"\\b\\d{1,2}(?:[\\.,]\\d*)?[NS]\\d{1,3}(?:[\\.,]\\d*)?[EW]\\b"];  // Must have a digit on the left side of the decimal
}

+ (NSString *) RegexAirports
{
    return [NSString stringWithFormat:@"((?:%@)|(?:@?\\b[A-Z0-9]{%d,%d}\\b))", [Airports RegAdHocFix], MIN(MIN_NAVAID_CODE_LENGTH, MIN_AIRPORT_CODE_LENGTH), MAX_AIRPORT_CODE_LENGTH];
}

static NSRegularExpression * _reAdhoc = nil;
static NSRegularExpression * _reAirports = nil;

+ (BOOL) isAdhocFix:(NSString *) sz
{
    NSError * err = NULL;
    if (_reAdhoc == nil)
        _reAdhoc = [[NSRegularExpression alloc] initWithPattern:[Airports RegAdHocFix] options:NSRegularExpressionCaseInsensitive error:&err];
    
    return [_reAdhoc numberOfMatchesInString:sz options:0 range:NSMakeRange(0, sz.length)];
}

- (instancetype)init
{
    self = [super init];
	if (self != nil)
	{
		self.rgAirports = [[NSMutableArray alloc] init];
		self.errorString = @"";
	}
	return self;
}

+ (MKCoordinateRegion) defaultRegionForPosition:(CLLocation *) loc
{
	double dlat = 1.0, dlon = 1.0;
	
	return MKCoordinateRegionMake(loc.coordinate, MKCoordinateSpanMake(dlat, dlon));
}

- (BOOL) loadAirportsNearPosition:(MKCoordinateRegion) loc limit:(NSInteger)max
{
	self.errorString = @"";
	
	if (!self.rgAirports)
		self.rgAirports = [[NSMutableArray alloc] init];
	[self.rgAirports removeAllObjects];
		
    MFBAppDelegate * app = MFBAppDelegate.threadSafeAppDelegate;
	
    LocalAirports * la = [[LocalAirports alloc] initWithLocation:loc withDB:app.db withLimit:max];
    if (la.rgAirports != nil)
        self.rgAirports = la.rgAirports;
    else
        self.errorString = NSLocalizedString(@"Local Airports initWithLocation found no airports.", "Error message for failure to find an airport in the db");

	return [self.errorString length] == 0;
}

+ (NSString *) appendAirport:(MFBWebServiceSvc_airport *)ap ToRoute:(NSString *) szRouteSoFar
{
	NSString * szReturn = szRouteSoFar;
    
    if (szReturn == nil || [szReturn length] == 0)
        return ap.Code;
	
	if (ap != nil && ap.Code != nil && [ap.Code length] > 0)
	{
		// check that this airport is not already at the end of the list
		NSString * szCurrent = szRouteSoFar;
		NSRange r = [szCurrent rangeOfString:ap.Code options:NSBackwardsSearch|NSCaseInsensitiveSearch];
		
		// if it's not at the end of the list OR szCurrent is still nil, append it.
		if (szCurrent == nil || r.location == NSNotFound || (r.location + r.length) < [szCurrent length])
		{
			NSString * newText = [NSString stringWithFormat:@"%@ %@", (szCurrent != nil) ? szCurrent : @"", ap.Code];
			szReturn = [newText uppercaseString];
		}
	}
	return szReturn;
}

+ (NSString *) appendNearestAirport:(NSString *) szRouteSoFar
{
	Airports * ap = [[Airports alloc] init];
	
	MFBAppDelegate * app = MFBAppDelegate.threadSafeAppDelegate;
	
	if (app.mfbloc.lastSeenLoc == nil)
	{
#ifdef DEBUG
		NSLog(@"No current loc in appendNearestAirport");
#endif
		return szRouteSoFar;
	}
	
	[ap loadAirportsNearPosition:[Airports defaultRegionForPosition:app.mfbloc.lastSeenLoc] limit:1];
	
	if ([ap.rgAirports count] > 0)
	{
		MFBWebServiceSvc_airport * airport = (MFBWebServiceSvc_airport *) (ap.rgAirports)[0];
		return [Airports appendAirport:airport ToRoute:szRouteSoFar];
	}
	else
		return szRouteSoFar;
}

- (BOOL) loadAirportsFromRoute:(NSString *)szRoute; 
{
	NSLog(@"loadAirportsFromRoute");
	self.errorString = @"";
	
	if (!self.rgAirports)
		self.rgAirports = [[NSMutableArray alloc] init];
	[self.rgAirports removeAllObjects];

	MFBAppDelegate * app = MFBAppDelegate.threadSafeAppDelegate;
	
    LocalAirports * la = [[LocalAirports alloc] initWithAirportList:szRoute withDB:app.db fromLoc:app.mfbloc.lastSeenLoc.coordinate];
    if (la.rgAirports == nil)
        self.errorString = NSLocalizedString(@"Unable to initialize airport list", @"Error message when looking up airports in a route of flight");
    else
        self.rgAirports = la.rgAirports;
	
	return [self.errorString length] == 0;
}

- (double) maxDistanceOnRoute:(NSString *) szRoute;
{
    double dist = 0.0;
    
    [self loadAirportsFromRoute:szRoute];
    
    NSInteger cAirports = [self.rgAirports count];
    int i, j;
    // find the distance between each of the airports.
    for (i = 0; i < cAirports; i++)
    {
        MFBWebServiceSvc_airport * ap1 = (MFBWebServiceSvc_airport *) (self.rgAirports)[i];
        
        if (![ap1 isPort])
            continue;
        
        for (j = i + 1; j < cAirports; j++)
        {
            MFBWebServiceSvc_airport * ap2 = (MFBWebServiceSvc_airport *) (self.rgAirports)[j];
            
            if (![ap2 isPort])
                continue;
            
            double lat1rad = DEG2RAD([ap1.Latitude doubleValue]);
            double lon1 = [ap1.Longitude doubleValue];
            double lat2rad = DEG2RAD([ap2.Latitude doubleValue]);
            double lon2 = [ap2.Longitude doubleValue];
            double d = acos(sin(lat1rad) * sin(lat2rad) + cos(lat1rad) * cos(lat2rad) * cos(DEG2RAD(lon2) - DEG2RAD(lon1))) * 3440.06479;
            if (d > dist)
                dist = d;
        }
    }
    
    return dist;
}

- (MKCoordinateRegion) defaultZoomRegionWithPath:(MFBWebServiceSvc_ArrayOfLatLong *) rgll
{
	LatLongBox * llb = [[LatLongBox alloc] init];
	
	CLLocationCoordinate2D loc;
	
	if ([self.rgAirports count] > 0)
	{
		for (MFBWebServiceSvc_airport * ap in self.rgAirports)
		{
			loc.latitude = [ap.Latitude doubleValue];
			loc.longitude = [ap.Longitude doubleValue];
			[llb addPoint:loc];
		}
	}
	
	if (rgll != nil && [rgll.LatLong count] > 0)
	{
		for (MFBWebServiceSvc_LatLong * ll in rgll.LatLong)
		{
			loc.latitude = [ll.Latitude doubleValue];
			loc.longitude = [ll.Longitude doubleValue];
			[llb addPoint:loc];
		}
	}

	return [llb getRegion];
}

+ (NSArray *) CodesFromString:(NSString *) szAirports
{
    NSMutableArray * rgResult = [NSMutableArray new];
    NSError * err = NULL;
    if (_reAirports == nil)
        _reAirports = [[NSRegularExpression alloc] initWithPattern:[Airports RegexAirports] options:NSRegularExpressionCaseInsensitive error:&err];
    szAirports = [szAirports uppercaseString];
    NSArray * rgMatches = [_reAirports matchesInString:szAirports options:0 range:NSMakeRange(0, szAirports.length)];
    for (NSTextCheckingResult * tcr in rgMatches)
        [rgResult addObject:[szAirports substringWithRange:tcr.range]];

    return rgResult;
}
@end

@implementation LatLongBox

- (instancetype)init
{
    self = [super init];
	if (self != nil)
	{
		latMin = latMax = longMin = longMax = 0.0;
		cPoints = 0;
	}
	return self;
}

- (BOOL) isValid
{
	return cPoints > 0;
}

- (BOOL) isInfinitessimal
{
	return (latMax == latMax || longMin == longMax);
}

- (void) addPoint:(CLLocationCoordinate2D)loc
{
	if (cPoints == 0) // 1st point
	{
		latMin = latMax = loc.latitude;
		longMin = longMax = loc.longitude;
	}
	else 
	{
		if (loc.latitude < latMin)
			latMin = loc.latitude;
		if (loc.latitude > latMax)
			latMax = loc.latitude;
		if (loc.longitude < longMin)
			longMin = loc.longitude;
		if (loc.longitude > longMax)
			longMax = loc.longitude;
	}

	cPoints++;
}

- (MKCoordinateRegion) getRegion
{
	MKCoordinateRegion mcr;
	mcr.span.latitudeDelta = mcr.span.longitudeDelta = 0.0;
	mcr.center.latitude = mcr.center.longitude = 0.0;

	if ([self isValid])
	{
		if ([self isInfinitessimal])
		{
			latMin -= 0.5;
			latMax += 0.5;
			longMin -= 0.5;
			longMax += 0.5;			
		}
		
		mcr.center.latitude = (latMax + latMin) / 2.0;
		mcr.center.longitude = (longMax + longMin) / 2.0;
		mcr.span.latitudeDelta = (latMax - latMin);
		mcr.span.longitudeDelta = (longMax - longMin);
	}

	return mcr;
}
@end

@implementation MFBWebServiceSvc_airport (Annotatable)
- (instancetype) initFromRow:(sqlite3_stmt *) row
{
    self.Code = @((char *)sqlite3_column_text(row, 0));
    self.Name = @((char *)sqlite3_column_text(row, 1));
    self.LatLong = [MFBWebServiceSvc_LatLong new];
    self.LatLong.Latitude = @(sqlite3_column_double(row, 4));
    self.LatLong.Longitude = @(sqlite3_column_double(row, 5));
    self.Latitude = self.LatLong.Latitude.stringValue;
    self.Longitude = self.LatLong.Longitude.stringValue;
    self.FacilityTypeCode = @((char *)sqlite3_column_text(row, 2));
    // column 3 is sourceusername - ignore it.
    // column 6 is preferred - ignore it
    const char * sz = (char *) sqlite3_column_text(row, 7);
    self.Country = (sz == NULL) ? @"" : @(sz);
    sz = (char *) sqlite3_column_text(row, 8);
    self.Admin1 = (sz == NULL) ? @"" : @(sz);
    self.DistanceFromPosition = @(sqlite3_column_double(row, 9));
    return self;
}

- (CLLocationCoordinate2D) coordinate
{
    if (self.LatLong != nil)
        return CLLocationCoordinate2DMake(self.LatLong.Latitude.doubleValue, self.LatLong.Longitude.doubleValue);
    else
        return CLLocationCoordinate2DMake(self.Latitude.doubleValue, self.Longitude.doubleValue);
}

- (NSString *) title
{
    double dist = [self.DistanceFromPosition doubleValue];
 
    MFBAppDelegate * app = MFBAppDelegate.threadSafeAppDelegate;

    if (dist == 0.0 && app.mfbloc.lastSeenLoc != nil)
    {
        // try to compute the distance
        CLLocationCoordinate2D cloc = app.mfbloc.lastSeenLoc.coordinate;
        if (cloc.latitude != 0.0 && cloc.longitude != 0.0)
            dist = NM_IN_A_METER * [app.mfbloc.lastSeenLoc distanceFromLocation:[[CLLocation alloc] initWithLatitude:self.Latitude.doubleValue longitude:self.Longitude.doubleValue]];
    }
    NSString * szDistance = [NSString localizedStringWithFormat:NSLocalizedString(@" (%.1fNM away)", @"Distance to airport - %.1f is replaced by the distance in nautical miles"), dist];
    
    return [NSString stringWithFormat:@"%@%@", self.Code, ((dist == 0.0) ? @"" : szDistance)];
}

- (NSComparisonResult) compareDistance:(MFBWebServiceSvc_airport *) ap
{
	if ([ap.DistanceFromPosition doubleValue] > [self.DistanceFromPosition doubleValue])
		return NSOrderedAscending;
	else if ([ap.DistanceFromPosition doubleValue] < [self.DistanceFromPosition doubleValue])
		return NSOrderedDescending;
	else
		return NSOrderedSame;

}

- (NSString *) subtitle
{
    NSString * szLocale = (Country == nil || Country.length == 0 || [Country hasPrefix:@"--"]) ? @"" :
        [[NSString stringWithFormat:@"%@%@", (Admin1 == nil || Admin1.length == 0) ? Country : [NSString stringWithFormat:@"%@, ", Admin1], Country] stringByTrimmingCharactersInSet:NSCharacterSet.whitespaceCharacterSet];
	
    return szLocale.length == 0 ? self.Name : [NSString stringWithFormat:@"%@ (%@)", self.Name, szLocale];
}

+ (MFBWebServiceSvc_airport *) getAdHoc:(NSString *) szLatLon;
{
    MFBWebServiceSvc_airport * ap = [MFBWebServiceSvc_airport new];
    MFBWebServiceSvc_LatLong * ll = [MFBWebServiceSvc_LatLong fromString:szLatLon];
    if (ll == nil)
        return nil;
    MFBLocation * loc = MFBAppDelegate.threadSafeAppDelegate.mfbloc;
    ap.LatLong = ll;
    ap.Latitude = ll.Latitude.stringValue;
    ap.Longitude = ll.Longitude.stringValue;
    ap.FacilityType = ap.FacilityTypeCode = @"FX";
    ap.DistanceFromPosition = @(NM_IN_A_METER * (loc.lastSeenLoc == nil ? 0 : [loc.lastSeenLoc distanceFromLocation:[[CLLocation alloc] initWithLatitude:ll.Latitude.doubleValue longitude:ll.Longitude.doubleValue]]));
    ap.Name = [ll description];
    ap.Code = szLatLon;
    ap.UserName = @"";
    return ap;
}

- (BOOL) isAdhoc
{
    return [Airports isAdhocFix:[szNavaidPrefix stringByAppendingString:self.Code]];
}

- (BOOL) isPort
{
    return ([self.FacilityTypeCode compare:@"A"] == NSOrderedSame || 
            [self.FacilityTypeCode compare:@"H"] == NSOrderedSame || 
            [self.FacilityTypeCode compare:@"S"] == NSOrderedSame);
}

// Returns a priority for different navaids in disambiguation; lower is higher priority
- (int) NavaidPriority
{
    // Airports ALWAYS have priority
    if ([self isPort])
        return 0;

    // Otherwise, give priority to VOR/VORTAC/etc., else NDB, else GPS fix, else everything else
    // VOR Types:
    if ([self.FacilityTypeCode compare:@"V"] == NSOrderedSame || 
        [self.FacilityTypeCode compare:@"C"] == NSOrderedSame || 
        [self.FacilityTypeCode compare:@"D"] == NSOrderedSame || 
        [self.FacilityTypeCode compare:@"T"] == NSOrderedSame)
        return 1;
    
    // NDB Types:
    if ([self.FacilityTypeCode compare:@"R"] == NSOrderedSame || 
        [self.FacilityTypeCode compare:@"RD"] == NSOrderedSame || 
        [self.FacilityTypeCode compare:@"M"] == NSOrderedSame || 
        [self.FacilityTypeCode compare:@"MD"] == NSOrderedSame || 
        [self.FacilityTypeCode compare:@"U"] == NSOrderedSame)
        return 2;
    
    // Generic fix
    if ([self.FacilityTypeCode compare:@"FX"] == NSOrderedSame)
        return 3;
    
    return 4;
}

- (NSString *) description
{
    return [NSString stringWithFormat:@"%@ (%@) %@", self.FacilityTypeCode, self.Code, self.Name];
}

@end

@implementation MFBWebServiceSvc_VisitedAirport(Sortable)
- (NSComparisonResult) compareName:(MFBWebServiceSvc_VisitedAirport *) va
{
    return [self.Airport.Name compare:va.Airport.Name options:NSCaseInsensitiveSearch];    
}

- (NSString *) AllCodes
{
    if (Aliases == nil)
        return Code;
    return [NSString stringWithFormat:@"%@,%@", Code, Aliases];
}

- (NSString *) description
{
    return self.Airport.description;
}

// Allow a visited airport to be annotatable based on the underlying airport.
- (CLLocationCoordinate2D) coordinate { return self.Airport.coordinate;}
- (NSString *) title {return self.Airport.title;}
- (NSString *) subtitle {return self.Airport.subtitle;}
@end

@implementation MFBWebServiceSvc_LatLong(Description)
- (NSString *) description
{
    NSNumberFormatter * nf = [[NSNumberFormatter alloc] init];
    nf.maximumFractionDigits = 8;
    return [NSString stringWithFormat:@"%@, %@", [nf stringFromNumber:self.Latitude], [nf stringFromNumber:self.Longitude]];
}

- (CLLocationCoordinate2D) coordinate
{
    return CLLocationCoordinate2DMake(self.Latitude.doubleValue, self.Longitude.doubleValue);
}

+ (MFBWebServiceSvc_LatLong *) fromString:(NSString *)sz
{
    NSError * err = NULL;
    sz = [sz uppercaseString];
    NSRegularExpression * reAirports = [[NSRegularExpression alloc] initWithPattern:@"@?([^a-zA-Z]+)([NS]) *([^a-zA-Z]+)([EW])" options:NSRegularExpressionCaseInsensitive error:&err];
    NSArray * rgMatches = [reAirports matchesInString:sz options:0 range:NSMakeRange(0, sz.length)];
    if (rgMatches.count > 0)
    {
        NSTextCheckingResult * tcr = rgMatches[0];
        @try {
            MFBWebServiceSvc_LatLong * ll = [MFBWebServiceSvc_LatLong new];
            NSString * szLatString = [sz substringWithRange:[tcr rangeAtIndex:1]];
            NSString * szLatNS = [sz substringWithRange:[tcr rangeAtIndex:2]];
            NSString * szLonString = [sz substringWithRange:[tcr rangeAtIndex:3]];
            NSString * szLonEW = [sz substringWithRange:[tcr rangeAtIndex:4]];
            ll.Latitude = [NSNumber numberWithDouble:[szLatString doubleValue] * (([szLatNS compare:@"N"] == NSOrderedSame) ? 1 : -1)];
            ll.Longitude = [NSNumber numberWithDouble:[szLonString doubleValue] * (([szLonEW compare:@"E"] == NSOrderedSame) ? 1 : -1)];
            return ll;
        }
        @catch (NSException * ex)
        {
            return nil;
        }
        return nil;
    }
    return nil;
}

- (instancetype) initWithCoord:(CLLocationCoordinate2D) coord
{
    self.Latitude = @(coord.latitude);
    self.Longitude = @(coord.longitude);
    return self;
}

- (NSString *) toAdhocString
{
    NSNumberFormatter * nf = [NSNumberFormatter new];
    nf.maximumFractionDigits = 4;

    return [NSString stringWithFormat:@"%@%@%@%@%@",
            szNavaidPrefix,
            [nf stringFromNumber:@(ABS(Latitude.doubleValue))],
            Latitude.doubleValue > 0 ? @"N" : @"S",
            [nf stringFromNumber:@(ABS(Longitude.doubleValue))],
            Longitude.doubleValue > 0 ? @"E" : @"W"];
}
@end

File: Classes/Model classes/Places/Airports.swift
/*
    MyFlightbook for iOS - provides native access to MyFlightbook
    pilot's logbook
 Copyright (C) 2023 MyFlightbook, LLC
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//
//  Airports.swift
//  MyFlightbook
//
//  Created by Eric Berman on 2/25/23.
//

import Foundation
import MapKit
import SQLite3

@objc public class Airports : NSObject {
    @objc public var errorString = ""
    @objc public var rgAirports : [MFBWebServiceSvc_airport] = []

    public static let szUSAirportPrefix = "K"
    public static let szNavaidPrefix = "@"
    
    private static let MIN_NAVAID_CODE_LENGTH = 2
    private static let MIN_AIRPORT_CODE_LENGTH = 3
    private static let MAX_AIRPORT_CODE_LENGTH = 6
        
    private static let RegAdHocFix = "\(szNavaidPrefix)\\b\\d{1,2}(?:[\\.,]\\d*)?[NS]\\d{1,3}(?:[\\.,]\\d*)?[EW]\\b"  // Must have a digit on the left side of the decimal
    
    private static let RegexAirports = String(format:"((?:%@)|(?:@?\\b[A-Z0-9]{%d,%d}\\b))", RegAdHocFix, min(MIN_NAVAID_CODE_LENGTH, MIN_AIRPORT_CODE_LENGTH), MAX_AIRPORT_CODE_LENGTH)
    
    private static var _reAdhoc : NSRegularExpression?
    private static var _reAirports : NSRegularExpression?
    
    public static func degreesToRadians(_ degrees : Double) -> Double {
        return (degrees * 0.0174532925199433)
    }
    
    public static func isAdhocFix(sz : String) -> Bool {
        if (_reAdhoc == nil) {
            _reAdhoc = try! NSRegularExpression(pattern: Airports.RegAdHocFix, options:.caseInsensitive)
        }
        return _reAdhoc!.numberOfMatches(in: sz, range: NSMakeRange(0, sz.count)) > 0
    }
    
    @objc public override init() {
        super.init()
    }
    
    @objc public static func defaultRegionForPosition(_ loc : CLLocation) -> MKCoordinateRegion {
        return MKCoordinateRegion(center: loc.coordinate, span: MKCoordinateSpan(latitudeDelta: 1.0, longitudeDelta: 1.0))
    }
    
    @discardableResult @objc(loadAirportsNearPosition: limit:) public func loadAirportsNearPosition(_ loc : MKCoordinateRegion, max : Int) -> Bool {
        errorString = ""
        rgAirports = []
        
        let db = SwiftHackBridge.getDB()

        let la = LocalAirports(loc: loc, db: db, limit: max)
        rgAirports = la.rgAirports
        return !errorString.isEmpty
    }
    
    @objc(appendAirport:ToRoute:) public static func appendAirport(_ ap : MFBWebServiceSvc_airport?, szRouteSoFar : String?) -> String {
        var szReturn = szRouteSoFar ?? ""

        if szReturn.isEmpty {
            return ap?.code ?? ""
        }
        
        if (!(ap?.code ?? "").isEmpty) {
            // check that this airport is not already at the end of the list
            let szCurrent = szReturn    // guaranteed non-nil
            let r = szCurrent.range(of: ap!.code!, options: [.backwards, .caseInsensitive])
 
            // if it's not at the end of the list OR szCurrent is still nil, append it.
            if (r == nil) {
                szReturn = "\(szReturn.trimmingCharacters(in: .whitespaces)) \(ap!.code!)".uppercased().trimmingCharacters(in: .whitespaces)
            }
        }
        return szReturn
    }
    
    @objc public static func appendNearestAirport(_ szRouteSoFar : String) -> String {
        let ap = Airports()
        let lastLoc = SwiftHackBridge.lastLoc()
        if (lastLoc == nil) {
            return szRouteSoFar
        }
        
        ap.loadAirportsNearPosition(Airports.defaultRegionForPosition(lastLoc!), max: 1)
        return (ap.rgAirports.isEmpty) ? szRouteSoFar : Airports.appendAirport(ap.rgAirports[0], szRouteSoFar: szRouteSoFar)
    }
    
    @objc public func loadAirportsFromRoute(_ szRoute : String) -> Void {
        NSLog("loadAirportsFromRoute")
        errorString = ""
        
        rgAirports = []
        
        let loc = SwiftHackBridge.lastLoc()
        
        let la = LocalAirports.init(szAirports: szRoute, db: SwiftHackBridge.getDB(), loc: loc?.coordinate)
        
        self.rgAirports = la.rgAirports
    }
    
    @objc public func maxDistanceOnRoute(_ szRoute : String) -> Double {
        var dist = 0.0
        
        self.loadAirportsFromRoute(szRoute)
        
        let cAirports = rgAirports.count
        for i in 0 ..< cAirports {
            let ap1 = rgAirports[i]
            if (!ap1.isPort()) {
                continue
            }
            
            for j in i + 1 ..< cAirports {
                let ap2 = rgAirports[j]
                
                if (!ap2.isPort()) {
                    continue
                }
                
                let lat1rad = Airports.degreesToRadians(ap1.latLong.latitude.doubleValue)
                let lon1 = ap1.latLong.longitude.doubleValue
                let lat2rad = Airports.degreesToRadians(ap2.latLong.latitude.doubleValue)
                let lon2 = ap2.latLong.longitude.doubleValue
                let d = acos(sin(lat1rad) * sin(lat2rad) + cos(lat1rad) * cos(lat2rad) * cos(Airports.degreesToRadians(lon2) - Airports.degreesToRadians(lon1))) * 3440.06479;
                dist = max(d, dist)
            }
        }
        return dist
    }
    
    @objc public func defaultZoomRegionWithPath(_ rgll : MFBWebServiceSvc_ArrayOfLatLong?) -> MKCoordinateRegion {
        let llb = LatLongBox()
        
        if (!rgAirports.isEmpty) {
            for ap in rgAirports {
                llb.addPoint(ap.latLong.coordinate())
            }
        }
        
        if ((rgll?.latLong?.count ?? 0) > 0) {
            for ll in rgll!.latLong {
                if let l = ll as? MFBWebServiceSvc_LatLong {
                    llb.addPoint(l.coordinate())
                }
            }
        }
        return llb.getRegion()
    }
    
    @objc public static func CodesFromString(_ szAirports : String) -> [String] {
        var rgResult : [String] = []
        if (_reAirports == nil) {
            _reAirports = try! NSRegularExpression(pattern: Airports.RegexAirports, options: .caseInsensitive)
        }
        
        let airports = szAirports.uppercased()
        let rgMatches = _reAirports!.matches(in: airports, range: NSMakeRange(0, airports.count))
        for tcr in rgMatches {
            rgResult.append(String(airports[Range(tcr.range, in: airports)!]))
        }

        return rgResult
    }
}

@objc class LocalAirports : NSObject {
    @objc public var database : OpaquePointer? = nil
    @objc public var rgAirports : [MFBWebServiceSvc_airport] = []
    
    private static func distanceColumnFromLoc(_ loc : CLLocationCoordinate2D?) -> String {
        if (loc == nil) {
            return "0 AS Distance"  // no loc = zero distance
        }
        
        return String(format: "distance(ap.latitude, ap.longitude, %.8f, %.8f) AS Distance",
                      loc!.latitude,
                      loc!.longitude)
    }
    
    private static func isUSAirport(_ szAirport : String) -> Bool {
        return szAirport.count == 4 && szAirport.hasPrefix(Airports.szUSAirportPrefix)
    }
    
    private static func USAirportPrefix(_ szAirport : String) -> String {
        return isUSAirport(szAirport) ? String(szAirport[szAirport.index(after: szAirport.startIndex)...]) : szAirport
    }
    
    private override init() {
        super.init()
    }
    
    @objc(initWithLocation: withDB: withLimit:) public convenience init(loc : MKCoordinateRegion, db : OpaquePointer?, limit : Int) {
        self.init()
        database = db
        rgAirports =  []
        
        // if the span is more than about 4 degrees in either direction, return
        // an empty set - we're too far zoomed out to be meaningful.
        if (loc.span.latitudeDelta > 4.0 || loc.span.longitudeDelta > 4.0) {
            return
        }
        
        var sqlAirportsNearPosition : OpaquePointer?
        
        let lat = loc.center.latitude
        let lon = loc.center.longitude
        
        // BUG: this doesn't work if we cross 180 degrees, but there are so few airports there that it shouldn't matter
        let minLat = max(lat - (loc.span.latitudeDelta / 2.0), -90.0);
        let maxLat = min(lat + (loc.span.latitudeDelta / 2.0), 90.0);
        let minLong = lon - (loc.span.longitudeDelta / 2.0);
        let maxLong = lon + (loc.span.longitudeDelta / 2.0);

        // we don't bother correcting lon's below -180 or above +180 for reason above
        let lastLoc = SwiftHackBridge.lastLoc()
        let curLoc = (lastLoc == nil) ? loc.center : lastLoc!.coordinate
        
        let fHeliports = UserPreferences.current.includeHeliports
        
        let szSql = String(format: "SELECT ap.*, %@ FROM airports ap WHERE ap.latitude BETWEEN %.8F AND %.8F AND ap.longitude BETWEEN %.8F AND %.8F AND Type IN %@ ORDER BY ROUND(Distance, 2) ASC, Preferred DESC, length(AirportID) DESC %@",
                           LocalAirports.distanceColumnFromLoc(curLoc),
                           minLat, maxLat, minLong, maxLong,
                           fHeliports ? "('H', 'A', 'S')" : "('A', 'S')",
                           limit > 0 ? String(format: "LIMIT %d", limit) : "")
        if (sqlite3_prepare(database, szSql.cString(using: .ascii), -1, &sqlAirportsNearPosition, nil) != SQLITE_OK) {
            NSLog("Error: failed to prepare statement with message '%s'.", sqlite3_errmsg(database))
        }
        
        while (sqlite3_step(sqlAirportsNearPosition) == SQLITE_ROW) {
            rgAirports.append(MFBWebServiceSvc_airport(row: sqlAirportsNearPosition))
        }
        
        sqlite3_finalize(sqlAirportsNearPosition)
    }
    
    public convenience init(szAirports : String, db : OpaquePointer?, loc : CLLocationCoordinate2D?) {
        self.init()
        database = db
        rgAirports = []
        
        var dictAp : [String : MFBWebServiceSvc_airport] = [:]
        var sqlResolveAirports : OpaquePointer?
        
        // Break up the string into constituent airports
        let rgCodes = Airports.CodesFromString(szAirports)
        if (rgCodes.isEmpty) {
            return
        }
        
        // Find any matches in the database.
        // Need to strip any leading "@" for the actual search; we'll apply this logic further below.
        let szAirportSet = NSMutableString()
        for ac in rgCodes {
            var airportCode = ac
            // If it's an ad-hoc fix, just add it directly to the dictionary
            if (Airports.isAdhocFix(sz: airportCode)) {
                let apAdHoc = MFBWebServiceSvc_airport.getAdHoc(airportCode)
                if (apAdHoc != nil) {
                    dictAp[airportCode] = apAdHoc
                    continue
                }
            }
            
            // Strip the leading "@" if necessary:
            if (airportCode .hasPrefix("@")) {
                airportCode = String(airportCode[airportCode.index(after: airportCode.startIndex)...])
            }
            szAirportSet.appendFormat("%@\"%@\"", szAirportSet.length > 0 ? ", " : "", airportCode)
            if (LocalAirports.isUSAirport(airportCode)) {
                szAirportSet.appendFormat("%@\"%@\"", szAirportSet.length > 0 ? ", " : "", LocalAirports.USAirportPrefix(airportCode))
            }
        }
        
        let szSql = String(format: "SELECT ap.*, %@ FROM airports ap WHERE ap.airportID in (%@)",
                           LocalAirports.distanceColumnFromLoc(loc), szAirportSet)
        
        if (sqlite3_prepare(database, szSql.cString(using: .ascii), -1, &sqlResolveAirports, nil) != SQLITE_OK) {
            NSLog("Error: failed to prepare statement with message '%s'.", sqlite3_errmsg(database))
        }
        
        // Read the results and put them into a dictionary by their code
        while (sqlite3_step(sqlResolveAirports) == SQLITE_ROW) {
            let ap = MFBWebServiceSvc_airport(row: sqlResolveAirports)
            
            // if it's an airport/heliport/seaport, store it using its code
            if (ap.isPort()) {
                dictAp[ap.code] = ap
            } else {
                // else store it using the navaid key (prefix "@"), replacing any lower priority (higher navaid priority) navaid
                let szKey = "\(Airports.szNavaidPrefix)\(ap.code!)"
                let ap2 = dictAp[szKey]
                if (ap2 == nil || ap.NavaidPriority() < ap2!.NavaidPriority()) {
                    dictAp[szKey] = ap
                }
            }
            // slightly more efficient than autorelease
        }
        sqlite3_finalize(sqlResolveAirports);
        
        // We now have in dictAp an dictionary of the typed airports, in no particular order and effectively deduped by the database
        // Now we need to add the airports to rgairports in the order in which they were typed.
        for szTypedCode in rgCodes {
            var ap = dictAp[szTypedCode]
            
            // if not found, see if it is under the navaid
            if (ap == nil) {
                ap = dictAp["\(Airports.szNavaidPrefix)\(szTypedCode)"]
            }
            
            // if that didn't work, try seeing if it's there without the "K" prefix
            if (ap == nil) {
                ap = dictAp[LocalAirports.USAirportPrefix(szTypedCode)]
            }
            
            if (ap != nil) {
                rgAirports.append(ap!)
            }
        }
    }
}
